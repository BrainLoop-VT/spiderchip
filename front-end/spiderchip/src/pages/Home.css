@import url('https://fonts.googleapis.com/css2?family=Press+Start+2P&display=swap');
@import url('https://fonts.cdnfonts.com/css/quantifier-nbp');
@import url('https://fonts.cdnfonts.com/css/perfect-dos-vga-437');

/* Ensure no scrolling or extra space */
html, body {
    margin: 0;
    padding: 0;
    overflow: hidden;
    width: 100%;
    height: 100%;
    box-sizing: border-box;
  }
  
  /* Center the page */
  .container {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    min-height: 100vh;
    width: 100vw;
    background-color: #6d6d6d;
    text-align: center;
  }
  
  /* Title text */
  .title {
    font-family: "Quantifier NBP", sans-serif;
    color: #ffffff;
    font-size: 7rem;
    margin-bottom: 3px;
  }
  
  /* Subtitle text */
  .subtitle {
    font-family: "Perfect DOS VGA 437", sans-serif;
    color: #d3d3d3;
    font-size: 1.7rem;
    margin-top: 5px;
    margin-bottom: 20px;
  }
  
  /* Start button */
  .startButton {
    font-family: "Press Start 2P", sans-serif;
    font-size: 1rem;
    padding: 6px 20px;
    border: 2px solid white;
    background-color: transparent;
    color: white;
    cursor: pointer;
    transition: 0.2s ease-in-out;
    text-align: center;
    min-width: 100px;
    margin-top: 10px;
  }
  
  .startButton:hover {
    background-color: white;
    color: black;
  }
  