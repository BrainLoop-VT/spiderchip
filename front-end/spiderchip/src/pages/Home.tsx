import './Home.css';
import { useState } from "react";
import AuthModal from "../components/AuthModal";
import { useNavigate } from 'react-router-dom';

const Home = () => {
  const [isModalOpen, setIsModalOpen] = useState(false);
  const navigate = useNavigate()

  const handleStartClick = () => {
    const token = localStorage.getItem("token");
    if (token) {
      navigate("/game");
    } else {
      setIsModalOpen(true);
    }
  };

  return (
    <div className="container">
      <h1 className="title">SPiderchip</h1>
      <p className="subtitle">an algorithms game</p>
      <button className="startButton" onClick={handleStartClick}>
        START
      </button>

      <AuthModal isOpen={isModalOpen} onClose={() => setIsModalOpen(false)} />
    </div>
  );
};

export default Home;