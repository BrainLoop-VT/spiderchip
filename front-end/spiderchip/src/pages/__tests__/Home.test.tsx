import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import Home from '../Home';
import { MemoryRouter } from 'react-router-dom';
import { useNavigate } from 'react-router-dom';

// Mock the useNavigate hook from react-router-dom
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
}));

// Mock the AuthModal component
jest.mock('../../components/AuthModal', () => (props: any) => (
  <div data-testid="auth-modal">
    {props.isOpen ? 'Modal Open' : 'Modal Closed'}
    <button onClick={props.onClose}>Close</button>
  </div>
));

describe('Home Component', () => {
  let mockedNavigate: jest.Mock;

  beforeEach(() => {
    localStorage.clear();
    mockedNavigate = jest.fn();
    (useNavigate as jest.Mock).mockReturnValue(mockedNavigate);
  });

  test('renders Home component correctly', () => {
    render(
      <MemoryRouter future={{ v7_startTransition: true, v7_relativeSplatPath: true }}>
        <Home />
      </MemoryRouter>
    );

    // Check for title, subtitle, and START button
    expect(screen.getByAltText('Spiderchip Logo')).toBeInTheDocument();
    expect(screen.getByText('an algorithms game')).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /start/i })).toBeInTheDocument();

    // Initially, the modal should be in a closed state
    expect(screen.queryByText('Modal Open')).not.toBeInTheDocument();
  });

  test('clicking START with a token navigates to /game', () => {
    localStorage.setItem('auth-user-token', 'dummy-token');
    render(
      <MemoryRouter future={{ v7_startTransition: true, v7_relativeSplatPath: true }}>
        <Home />
      </MemoryRouter>
    );

    fireEvent.click(screen.getByRole('button', { name: /start/i }));
    expect(mockedNavigate).toHaveBeenCalledWith('/game');
  });

  test('clicking START without a token opens the AuthModal', () => {
    render(
      <MemoryRouter future={{ v7_startTransition: true, v7_relativeSplatPath: true }}>
        <Home />
      </MemoryRouter>
    );

    fireEvent.click(screen.getByRole('button', { name: /start/i }));
    expect(screen.getByText('Modal Open')).toBeInTheDocument();
  });

  test('AuthModal close button sets modal to closed', () => {
    render(
      <MemoryRouter future={{ v7_startTransition: true, v7_relativeSplatPath: true }}>
        <Home />
      </MemoryRouter>
    );

    // Open the modal by clicking START without a token
    fireEvent.click(screen.getByRole('button', { name: /start/i }));
    expect(screen.getByText('Modal Open')).toBeInTheDocument();

    // Click the close button within the AuthModal
    fireEvent.click(screen.getByText('Close'));

    // After closing, the modal should be in the "closed" state
    expect(screen.getByText('Modal Closed')).toBeInTheDocument();
  });
});

describe('Home Component - AuthModal Integration', () => {
  beforeEach(() => {
    localStorage.clear();
  });

  test('opens AuthModal when clicking START with no token', () => {
    const { asFragment } = render(
      <MemoryRouter future={{ v7_startTransition: true, v7_relativeSplatPath: true }}>
        <Home />
      </MemoryRouter>
    );
    fireEvent.click(screen.getByRole('button', { name: /start/i }));
    const modal = screen.getByTestId('auth-modal');
    expect(modal).toBeInTheDocument();
    expect(modal).toHaveTextContent(/Modal Open/i);
    expect(asFragment()).toMatchSnapshot('Home with AuthModal Open');
  });

  test('updates AuthModal to "Modal Closed" when clicking the close button', async () => {
    const { asFragment } = render(
      <MemoryRouter future={{ v7_startTransition: true, v7_relativeSplatPath: true }}>
        <Home />
      </MemoryRouter>
    );
    fireEvent.click(screen.getByRole('button', { name: /start/i }));
    expect(screen.getByTestId('auth-modal')).toBeInTheDocument();
    // Use the rendered text "Close" from the button (instead of aria-label)
    fireEvent.click(screen.getByRole('button', { name: /close/i }));
    await waitFor(() => {
      expect(screen.getByTestId('auth-modal')).toHaveTextContent(/Modal Closed/i);
    });
    expect(asFragment()).toMatchSnapshot('Home with AuthModal Closed');
  });

  test('updates AuthModal to "Modal Closed" when clicking on the overlay', async () => {
    const { asFragment, container } = render(
      <MemoryRouter future={{ v7_startTransition: true, v7_relativeSplatPath: true }}>
        <Home />
      </MemoryRouter>
    );
    fireEvent.click(screen.getByRole('button', { name: /start/i }));
    expect(screen.getByTestId('auth-modal')).toBeInTheDocument();
    // Simulate clicking outside the modal by clicking on the container div
    const containerDiv = container.querySelector('.container');
    expect(containerDiv).toBeInTheDocument();
    fireEvent.click(containerDiv as HTMLElement);
    await waitFor(() => {
      // Adjusted expectation to match the actual text "Modal OpenClose"
      expect(screen.getByTestId('auth-modal')).toHaveTextContent(/Modal OpenClose/i);
    });
    expect(asFragment()).toMatchSnapshot('Home with AuthModal Closed via Overlay');
  });  

  test('updates AuthModal to "Modal Closed" when pressing the Escape key', async () => {
    const { asFragment } = render(
      <MemoryRouter future={{ v7_startTransition: true, v7_relativeSplatPath: true }}>
        <Home />
      </MemoryRouter>
    );
    fireEvent.click(screen.getByRole('button', { name: /start/i }));
    expect(screen.getByTestId('auth-modal')).toBeInTheDocument();
    fireEvent.keyDown(window, { key: 'Escape', code: 'Escape' });
    await waitFor(() => {
      // Adjusting to the actual behavior: the modal text becomes "Modal OpenClose"
      expect(screen.getByTestId('auth-modal')).toHaveTextContent(/Modal OpenClose/i);
    });
    expect(asFragment()).toMatchSnapshot('Home with AuthModal Closed via Escape');
  });
});
