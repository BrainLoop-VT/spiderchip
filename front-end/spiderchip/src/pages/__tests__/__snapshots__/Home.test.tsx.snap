// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component - AuthModal Integration opens AuthModal when clicking START with no token: Home with AuthModal Open 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <img
      alt="Spiderchip Logo"
      class="logo"
      src="test-file-stub"
    />
    <p
      class="subtitle"
    >
      an algorithms game
    </p>
    <button
      class="startButton"
    >
      START
    </button>
    <div
      data-testid="auth-modal"
    >
      Modal Open
      <button>
        Close
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Home Component - AuthModal Integration updates AuthModal to "Modal Closed" when clicking on the overlay: Home with AuthModal Closed via Overlay 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <img
      alt="Spiderchip Logo"
      class="logo"
      src="test-file-stub"
    />
    <p
      class="subtitle"
    >
      an algorithms game
    </p>
    <button
      class="startButton"
    >
      START
    </button>
    <div
      data-testid="auth-modal"
    >
      Modal Open
      <button>
        Close
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Home Component - AuthModal Integration updates AuthModal to "Modal Closed" when clicking the close button: Home with AuthModal Closed 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <img
      alt="Spiderchip Logo"
      class="logo"
      src="test-file-stub"
    />
    <p
      class="subtitle"
    >
      an algorithms game
    </p>
    <button
      class="startButton"
    >
      START
    </button>
    <div
      data-testid="auth-modal"
    >
      Modal Closed
      <button>
        Close
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Home Component - AuthModal Integration updates AuthModal to "Modal Closed" when pressing the Escape key: Home with AuthModal Closed via Escape 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <img
      alt="Spiderchip Logo"
      class="logo"
      src="test-file-stub"
    />
    <p
      class="subtitle"
    >
      an algorithms game
    </p>
    <button
      class="startButton"
    >
      START
    </button>
    <div
      data-testid="auth-modal"
    >
      Modal Open
      <button>
        Close
      </button>
    </div>
  </div>
</DocumentFragment>
`;
